inc/blocks/namespace.php

```diff
@@ -157,6 +159,86 @@ function on_save_post( int $post_ID, WP_Post $post, bool $update ) : void {
 	}
 }
 
+/**
+ * Synchronise any XBs on the widget page with a shadow post type.
+ *
+ * @param instance $instance Widget Object.
+ * @param new_instance $new_instance Widget Object.
+ * @param old_instance $old_instance Widget Object
+ * @param widget $widget Current Widget Instance
+ * @return object
+ */
+function on_widgets_save( $instance, $new_instance, $old_instance, $widget ) {
+
+	// Scan for XBs in the post content.
+	$blocks = parse_blocks( $new_instance['content'] );
+	$xbs = find_xbs( $blocks );
+
+	// Find referenced XBs.
+	$existing_posts = new WP_Query( [
+		'post_type' => POST_TYPE,
+		'post_status' => 'any',
+		'post_name__in' => array_filter( array_map( function ( array $xb ) : string {
+			return $xb['attrs']['clientId'];
+		}, $xbs ) ),
+		'posts_per_page' => 100,
+		'no_found_rows' => true,
+	] );
+
+	// Update or create XB posts.
+	foreach ( $xbs as $index => $xb ) {
+		if ( empty( $xb['attrs']['clientId'] ) ) {
+			continue;
+		}
+
+		// Extract existing post if found.
+		$posts = wp_list_filter( $existing_posts->posts, [
+			'post_name' => $xb['attrs']['clientId'],
+		] );
+		$posts = array_values( $posts ); // Reset keys.
+
+		// Generate a default using the current post title and instance number in the content.
+		$default_title = sprintf( 'Widget A/B Test (XB %s)', $widget->number );
+		if ( ! isset( $xb['attrs']['title'] ) ) {
+			$xb['attrs']['title'] = '';
+		}
+
+		if ( empty( $posts ) ) {
+			// Create new shadow XB post.
+			$xb_post_id = wp_insert_post( [
+				'post_type' => POST_TYPE,
+				'post_status' => 'publish',
+				'post_content' => serialize_block( $xb ),
+				'post_name' => $xb['attrs']['clientId'],
+				'post_author' => get_current_user_id(),
+				'post_title' => $xb['attrs']['title'] ?: $default_title,
+			] );
+			// Store XB type.
+			$xb_name = ( isset( $xb['name'] ) ? $xb['name'] : 'XB' );
+			update_post_meta( $xb_post_id, '_xb_type', $xb_name );
+			update_post_meta( $xb_post_id, '_xb_type_' . sanitize_key( $xb_name ), true );
+		} else {
+			// Update existing post.
+			$xb_post_id = $posts[0]->ID;
+			wp_update_post( [
+				'ID' => $posts[0]->ID,
+				'post_content' => serialize_block( $xb ),
+				'post_title' => $xb['attrs']['title'] ?: $default_title,
+			] );
+		}
+
```

Comment:
Lets set some post meta so we know this XB is a widget, maybe with the key `_xb_is_widget`

Reply:
Should it not be set here? https://github.com/humanmade/aws-analytics/blob/972aea11d28fffb3dc1fea3cf4982c0dc6dea86b/inc/blocks/namespace.php#L139-L140

Reply:
Sorry I thought I'd mentioned rebasing this PR against `master` - if you look at the latest code you'll see it has changed a bit.

If you're not familiar with it rebasing in git means to replay all of your commits over another branch - most often you'd do this to make sure your branch will merge cleanly and that you have the latest code. After making sure your primary branch is up to date, from your working branch you would run `git rebase master`.

You may run into conflicts but the output should tell you how to go about resolving these.